### ISSUE TITLE
floorplan - icon positioning limited to square area?

### ISSUE THREAD

**Issue by flatsiedatsie** [mozilla-iot/gateway#2522]
I cannot drag the icons on the floorplan page over the entire page. They seem to be limited to a square area in the center?

**Comment by benfrancis** [comment]
That is correct, it's due to the scaling of the image on different screen sizes and aspect ratios. If you placed an icon outside the floorplan image in landscape, that space not exist when the page is displayed in portrait, and vice-versa.

![Screen Shot 2020-06-18 at 11 04 47](https://user-images.githubusercontent.com/552417/85007529-8fbab000-b153-11ea-80c9-94cf24ba3ff4.png)
![Screen Shot 2020-06-18 at 11 04 50](https://user-images.githubusercontent.com/552417/85007523-8df0ec80-b153-11ea-984e-5ab49238c84a.png)

**Comment by flatsiedatsie** [comment]
I figured as much. 

Still, it's not reallty clear to the end user. When I first tried to move an icon to the left and it just 'dropped' all of a sudden I didn't understand what was going on. Only when I had moved multiple icons I started to see the 'invisible' wall. 

It's also a bit of a waste of precious space on smaller screens.

Could there perhaps be a way to favour horizontal screens? Mobile phone and tablet users could then rotate their devices to edit optimally. Or perhaps the dimensions of the uploaded image would form the boundary?

**Comment by benfrancis** [comment]
> Could there perhaps be a way to favour horizontal screens?

Yes that could be done by making the square a rectangle, but I don't think that's a good idea. Smartphones are more commonly used in portrait rather than landscape mode and if a user uploads a floorplan image that is taller than it is wide, the image may appear very small even if space is available to display the image on the page. The square was the best compromise I could come up with, but let me know if you come up with a better solution.

>  Or perhaps the dimensions of the uploaded image would form the boundary?

What happens to all of the saved co-ordinates if the user changes the floorplan image? This is the reason the square exists.

**Comment by benfrancis** [comment]
Idea: We could add some subtle UI hints like having an icon bounce of the boundary of the square when dragged near it to show that there is a boundary, or even showing a faint dotted line where the boundary lies.

**Comment by flatsiedatsie** [comment]
I understand, but it's such a waste of space on a small 1024x600 display.

> if a user uploads a floorplan image that is taller than it is wide, the image may appear very small even if space is available to display the image on the page

Good point. Forget the 'favour horizontal' concept :-).

If the uploaded image is the basis, then users could actually have a better interaction on their mobile phones if they uploaded a vertical map? Or if it's a horizontal map, they would have a better interaction if they rotate their phones. The square shape compromises useable space in either case.

> What happens to all of the saved co-ordinates if the user changes the floorplan image?

How about: if a user uploads a new floorplan image..
- If it has the same dimensions, there's no issue.
- If it has different dimensions, all the icons could just be reset to the initial position? If the image has different dimensions, it wil mean the icons are not longer positioned correctly anyway? (I'm not sure, I haven't been able to actually upload an image).

**Comment by flatsiedatsie** [comment]
> if a user uploads a floorplan image that is taller than it is wide, the image may appear very small even if space is available to display the image on the page

It seems that if you upload a map image that isn't square, you'll lose most of the available space anyway currently. It has to be contained inside the square, which has to be contained inside the browser. So you lose a lot of real estate (pun intended).

**Comment by flatsiedatsie** [comment]
As a side note, once you have about 20 devices, the floorplan gets a little swamped even on a desktop browser.
